pkg/kv/store_message.go

```diff
@@ -14,54 +13,46 @@ type StoreMessage struct {
 }
 
 func (s *StoreMessage) GetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
-	val, err := s.Store.Get(ctx, []byte(path))
+	v, err := s.Store.Get(ctx, []byte(path))
 	if err != nil {
-		return fmt.Errorf("failed on Get (path: %s): %w", path, err)
+		return err
 	}
-	return proto.Unmarshal(val, msg)
+	return proto.Unmarshal(v, msg)
 }
 
-func (s *StoreMessage) SetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
-	val, err := proto.Marshal(msg)
+func (s *StoreMessage) GetMsgPredicate(ctx context.Context, path string, msg protoreflect.ProtoMessage) (Predicate, error) {
+	vp, err := s.Store.GetValuePredicate(ctx, []byte(path))
 	if err != nil {
-		return fmt.Errorf("failed on Set (path: %s): %w", path, err)
+		return nil, err
 	}
-	return s.Store.Set(ctx, []byte(path), val)
+	if msg == nil {
+		return vp.Predicate, nil
+	}
+	err = proto.Unmarshal(vp.Value, msg)
+	if err != nil {
+		return nil, err
+	}
+	return vp.Predicate, nil
 }
 
-func (s *StoreMessage) SetIf(ctx context.Context, path string, msg protoreflect.ProtoMessage, pred protoreflect.ProtoMessage) error {
+func (s *StoreMessage) SetMsg(ctx context.Context, path string, msg protoreflect.ProtoMessage) error {
 	val, err := proto.Marshal(msg)
 	if err != nil {
-		return fmt.Errorf("failed on Marshal (path: %s): %w", path, err)
-	}
-	if pred == nil {
-		return s.Store.SetIf(ctx, []byte(path), val, nil)
+		return err
```

Comment:
Same as previous comment

Reply:
Will address it as part of Store implementation

Reply:
I wonder why you have decided to move it to the store implementation, as this will require all stores to make sure to log the error with the key. When done in the store_message level you only have to take care of it once WDYT?

Reply:
Wanted to get the infomation also when you directly access the storage.
- We can guide the storage implementor
- Create a wrapper that will wrap the underlying storage to enritch the return errors
WDYT?